### PR TITLE
Devel/buffer refactor

### PR DESCRIPTION
**Pull request by nobrowser**
* What does this PR do?

Buffer struct refactor as planned

* Are there points in the code the reviewer needs to double check?

If it merges, it should be fine.  Only shuffling code around has been done.

* Does this PR meet the acceptance criteria?

Well, not really - I just shuffled it around.  Will be another pass if needed.

* What are the relevant issue numbers?
#377 

Edited: Tidied (@flatcap)

### CODE REVIEW

**Comment by flatcap** [comment]
@neomutt/reviewers Please review

**Comment by gahr** [comment]
Should we try to keep .h/.c consistency? Specifically, should anything declared in myvar.h be defined in myvar.c?

**Comment by gahr** [comment]
Also, I am not sure `mutt_extract_token` really belongs to `buffer.(c|h)`.. I don't have a strong opinion.. 

other than that, good to go for me!

**Comment by flatcap** [comment]
> Should we try to keep .h/.c consistency?

Yes.

> Specifically, should anything declared in myvar.h be defined in myvar.c?

Ideally, yes, but myvar.h is a bit of a mixed bag.
That probably needs more rearranging.

> I am not sure mutt_extract_token really belongs to buffer.(c|h)

Hmm... It's slightly more BUFFER-y than not.

For now, I'm happy with this experiment.
I'll add the above to the list of what should considered to be broken out next.

**Comment by nobrowser** [comment]
I also considered creating a myvar.c, but note that the issue said *minimal* set of changes to to isolate `BUFFER` ... so I resisted that urge.  Besides, it may have required *further* splitting if, for example, something in init.c depended on the new myvar.c .

File: buffer.c [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+{
+  mutt_buffer_add (buf, s, mutt_strlen (s));
+}
+
+void mutt_buffer_addch (BUFFER* buf, char c)
+{
+  mutt_buffer_add (buf, &c, 1);
+}
+
+int mutt_extract_token (BUFFER *dest, BUFFER *tok, int flags)
```
Comment (nobrowser) [inline buffer.c]: This is line 145.  What to reindent here?

**Comment by flatcap** [comment]
@P-EB 
This change is **just** a reorganisation of functions, so I'm not sure it could be split meaningfully.
Yes, the code's a mess, but that's not @nobrowser's fault :-)

When "The Big Tidy<sup>™️</sup>" occurs, most of the problems will disappear.

I'm happy with the code as it is, for now.

**Comment by nobrowser** [comment]
I should like to point out that the root of all evil is protos.h.

It's that file that needs to be split more than any other.
The reason I created myvar.h and filter.h was to avoid an `#include "protos.h"` in the new buffer.c file.

**Comment by flatcap** [comment]
Yes, I was referring to the formatting problems in `buffer.[ch]`.

I'm open to suggestions about what you'd like to split out next?

**Comment by P-EB** [comment]
@flatcap the reindent issue seems to be introduced by the commit. I might be wrong but my diff seems to say so.

**Comment by flatcap** [comment]
@P-EB In GitHub's diff view the indent **looks** awful because the tabs aren't handled well, but adding `?ts=8` to the URL makes it look ok.

**Pull request merged**